feat(docs): document CI/CD workflow (#1742)
hashicorp/terraform-cdk: website/docs/cdktf/create-and-deploy/deployment-patterns.mdx

---
page_title: Deployment Patterns
description: >-
  Continuously deploy your applications with CDKTF.
---

# Deployment Patterns

You can configure your CI/CD environment to automatically deploy your CDKTF applications. This page explains different types of deployment patterns to help you choose the methods that are most appropriate for your use case.

## Deployment Methods

> You can provision your infrastructure using either CDKTF CLI commands (e.g., `cdktf deploy`) or by synthesizing your application into a Terraform configuration file and using Terraform directly. Both approaches allow you to use Terraform Cloud and integrate with a Continous Integration (CI) pipeline.

Reviewer: ```suggestion
You can provision your infrastructure using either CDKTF CLI commands (e.g., `cdktf deploy`) or by synthesizing your application into a Terraform configuration file and using Terraform directly. Both approaches allow you to use Terraform Cloud and integrate with a CI/CD pipeline.
```

- **Your application contains stacks with complex dependencies.** Each [stack](/cdktf/concepts/stacks) in a CDKTF application is like a separate Terraform working directory. Terraform respects complex dependencies within the same configuration, but it is not designed to manage complex dependencies between multiple working directories. For example, if you define a stack A that depends on data from stack B and stack C, Terraform runs will fail. This is because Terraform will not understand that it must wait to build stack A until both stack B and stack C are completed.

### Run `cdktf synth` and Use Terraform Directly

We recommend synthesizing your CDKTF application and using Terraform directly when you need to integrate into existing Terraform workflows. Your organization may already use Terraform Cloud or have integrated Terraform into CI pipelines. In these cases, you may want to run `cdktf synth` and set the output folder to a particular directory that you can reference in your existing workflows.

## Using Terraform Cloud

> Terraform Cloud lets you review the current state of your infrastructure and run progress in the Terraform Cloud UI. You can also use Terraform Cloud as the source of secrets via [workspace variables](/cloud-docs/workspaces/variables). If you use the local execution mode or any other backend you must specify the credentials for your providers locally through environment variables.

Reviewer: ```suggestion
Terraform Cloud lets you review the current state of your infrastructure and run progress in the Terraform Cloud UI. You can also use Terraform Cloud as the source of secrets by defining [workspace variables](/cloud-docs/workspaces/variables). You can only use workspace variables for secrets if you set your workspace to [remote execution mode](/cloud-docs/workspaces/settings#execution-mode), which means that Terraform runs are executed within Terraform Cloud. If you use the local execution mode or any other backend, you must specify the credentials for your providers locally through environment variables. 
```

> To use Terraform Cloud with a [CLI-driven Run Workflow](/cloud-docs/run/cli), you must enable [remote execution](/cloud-docs/run#remote-operations) in your Terraform Cloud workspace.

Reviewer: ```suggestion
Using Terraform Cloud with a [CLI-driven Run Workflow](/cloud-docs/run/cli), also requires setting your workspace to remote execution mode. 
```

~> **Warning:** Remote execution with non-interactive workflows requires auto-approved deployments. Minimize risk of unpredictable infrastructure changes and configuration drift by making sure that no one can change your infrastructure outside of your automated build pipeline. Refer to [Non-Interactive Workflows](/cloud-docs/run/cli#non-interactive-workflows) for details.

> To run Terraform Cloud with a [VCS-Driven Run Workflow](/cloud-docs/run/ui) you must check in your synthesized code.

Reviewer: Do we need to say this bit about VCS? i'm wondering if that is implied?

Author: What do you mean? Like what VCS is?

Reviewer: oh no i was just wondering if we needed to tell folks that they have to check their synthesized code in if they're using VCS. If you want to keep this in I'm okay with that, but I was also wondering if we NEED to tell people that or if folks will already know that you have to check things in to use VCS. Sorry, I was not clear there at all!

Author: I'd rather error on the side of being overly explicit here

Reviewer: okay I support that then :)

## Managing Secrets and Sensitive Data

> No matter if you manage your secrets through Terraform Cloud (in `remote-execution` mode) or through your normal CI, secrets should never appear in your synthesized code. [This is why you should read Secrets with Terraform Variables](/cdktf/create-and-deploy/best-practices#read-secrets-with-terraform-variables).

Reviewer: ```suggestion
You should [read secrets with Terraform Variables](/cdktf/create-and-deploy/best-practices#read-secrets-with-terraform-variables) so that they do not appear in your synthesized CDKTF code.
```

> ## Example Workflows

Reviewer: We just don't like to have two headers in a row with nothing in between :-) 
```suggestion
## Example Workflows

The following examples demonstrate common deployment workflows for your CDKTF application. 
```

### Deploy with Terraform and Terraform Cloud

You could create a workflow with the following steps:

1. Add a pre-commit step that runs `cdktf synth`. This creates Terraform configuration files under `cdktf.out/stacks/<stack-name>`.
2. Configure the VCS integration for a Terraform cloud workspace to point to the directory with the synthesized configuration files. To deploy multiple independent stacks, configure one Terraform Cloud workspace to point to the output directory for each stack.

Refer to [UI/VCS-Driven Run Workflow](/cloud-docs/run/ui) in the Terraform Cloud documentation for details about how to trigger Terraform runs from merges and commits.

### Terraform CDK GitHub Action

> If you are using GitHub Actions and you want a default review / deployment workflow, you can use the [Terraform CDK GitHub Action](https://github.com/marketplace/actions/terraform-cdk-action). It provides you with comments on your PRs and a default review workflow. If you need more customization, please take a look at the next example.

Reviewer: ```suggestion
You can use the [Terraform CDK GitHub Action](https://github.com/marketplace/actions/terraform-cdk-action) if you are using GitHub Actions and you want a default review / deployment workflow. The Terraform CDK GitHub Action comments Terraform plans on your pull requests and provides a default review workflow. If you need more customization, refer to [GitHub Actions CI and Terraform Cloud](#github-actions-ci-and-terraform-cloud).
```

> 

Reviewer: Can you say here if folks are synthesizing or running with CDKTF directly? I just wasn't sure I think because it's TFC.

### GitHub Actions CI and Terraform Cloud

Set the `TERRAFORM_CLOUD_TOKEN` environment variable to the [Terraform Cloud API token](https://www.terraform.io/cloud-docs/users-teams-organizations/api-tokens). This allows the CDKTF stacks `RemoteBackend` to connect to Terraform Cloud.

You must create a job like this for every stack so that you can pass the stack name into `cdktf diff <stack name>`. None of the stacks can depend on each other because Terraform cannot reason about which stacks must be launched first to respect these dependencies.

> The following example shows a GitHub Actions configuration where the CDKTF application is deployed from the `main` branch and uses Terraform Cloud. This file could be stored in `.github/workflows/cdktf-deploy.yml`.

Reviewer: ```suggestion
The following example shows a GitHub Actions configuration where the CDKTF application is deployed from the `main` branch and uses Terraform Cloud. You could store this file in `.github/workflows/cdktf-deploy.yml`.
```

        continue-on-error: true # We want to see the plan output even if the plan fails

      # Replace this step with any workflow you like, e.g. posting a message in a slack channel.
      - uses: actions/github-script@0.9.0
        if: github.event_name == 'pull_request'
        env:
          PLAN: "terraform\n${{ steps.plan.outputs.stdout }}" # CDKTF gets the output from the last step
        with:
          github-token: ${{ secrets.GH_COMMENT_TOKEN }} # Create a secret with a GitHub token that can comment on PRs
          script: |
            const output = `#### Terraform Plan 📖\`${{ steps.plan.outcome }}\`
            <details><summary>Show Plan</summary>
            \`\`\`${process.env.PLAN}\`\`\`
            </details>
            *Pusher: @${{ github.actor }}, Action: \`${{ github.event_name }}\`, Working Directory: \`${{ env.tf_actions_working_dir }}\`, Workflow: \`${{ github.workflow }}\`*`;
            github.issues.createComment({
              issue_number: context.issue.number,
              owner: context.repo.owner,
              repo: context.repo.repo,
              body: output
            })
```

### Running the CDKTF CLI on a general purpose CI

> If you have a general purpose CI like GitHub Actions or CircleCI configured, use the `cdktf deploy --auto-approve <your stacks>` command to trigger a deployment. This deploys your stacks in the right order and in parallel. To do set up this workflow, you must install the following software:

Reviewer: ```suggestion
If you have a general purpose CI like GitHub Actions or CircleCI configured, use the `cdktf deploy --auto-approve <your stacks>` command to trigger a deployment. This deploys your stacks in the right order and in parallel. To set up this workflow, you must install the following software:
```

- Node in version 14.0+
- The `cdktf-cli` CLI
- Terraform v1.0+
- Any other languages you use for your CDKTF application

~> **Warning:** Auto-approved deployments skip the option to review planned changes to your infrastructure before Terraform applies them. Minimize risk of unpredictable infrastructure changes and configuration drift by making sure that no one can change your infrastructure outside of your automated build pipeline. Refer to [Non-Interactive Workflows](/cloud-docs/run/cli#non-interactive-workflows) for details.

### Contribute Additional Examples

> If you want to share another solution on this page, please [open a GitHub issue](https://github.com/hashicorp/terraform-cdk/issues/new?assignees=&labels=enhancement%2C+new&template=feature-request.md&title=) or submit a pull request for this page.

Reviewer: ```suggestion
If you want to share another solution, please [open a GitHub issue](https://github.com/hashicorp/terraform-cdk/issues/new?assignees=&labels=enhancement%2C+new&template=feature-request.md&title=) or submit a pull request for this documentation page.
```

```ts
const adminPassword = new TerraformVariable(this, "adminPassword", {
  type: "string",
  description: "Admin password for the instance",
  sensitive: true,
});
new MyResource(this, "hello", {
  adminPassword: adminPassword.value, // use this instead of process.env.ADMIN_PASSWORD
});
```

In the example above if we had used `adminPassword: process.env.ADMIN_PASSWORD` the admin password environment variable set at synthesization time would have been written to the `cdk.tf.json` file for your stack. This would have been a security risk.

In contrast the `TerraformVariable` construct synthesizes a Terraform input. This input is not written to the `cdk.tf.json` file, so it is not a security risk. Instead it's only used at execution time by Terraform directly.

To pass a [Terraform variable through environment variables](/cli/config/environment-variables#tf_var_name) you would need to set `TF_VAR_adminPasword='<your password>'` in the execution environment.

## Example Workflows

### Terraform Cloud

> This workflow is an example for the 'Run `cdktf synth` and handle the deployment seperately' method.
> 
> Add a pre-commit step that runs `cdktf synth`. This creates Terraform configuration files under `cdktf.out/stacks/<stack-name>`. Then, configure the VCS integration for a Terraform cloud workspace to point to the appropriate path. You can use the same technique to deploy multiple independent stacks by configuring a Terraform Cloud workspace for each stack with the same VCS configuration, but a different target path.
> 
> Refer to [UI/VCS-Driven Run Workflow](/cloud-docs/run/ui) in the Terraform Cloud documentation for details.

Reviewer: ```suggestion
You could create a workflow with the following steps:
1. Add a pre-commit step that runs `cdktf synth`. This creates Terraform configuration files under `cdktf.out/stacks/<stack-name>`. 
1. Configure the VCS integration for a Terraform cloud workspace to point to the directory with the synthesized configuration files. To deploy multiple independent stacks, configure one Terraform Cloud workspace to point to the output directory for each stack.

Refer to [UI/VCS-Driven Run Workflow](/cloud-docs/run/ui) in the Terraform Cloud documentation for details about how to trigger Terraform runs from merges and commits.
```

> ### GitHub Actions CI and Terraform Cloud

Reviewer: Question. Is this another example of synthesizing and then using an external deployment method with Terraform Directly?

I'm trying to think through how we can connect the titles of these examples back to the options we present above.
```suggestion
### GitHub Actions CI and Terraform Cloud
```

> Set the `TERRAFORM_CLOUD_TOKEN` environment variable to the [Terraform Cloud API token](https://www.terraform.io/cloud-docs/users-teams-organizations/api-tokens). This allows the CDKTF stacks `RemoteBackend` to connect to Terraform Cloud.
> 
> The following example shows a GitHub Actions configuration where the CDKTF application is deployed from the `main` branch and uses Terraform Cloud. This file could be stored in `.github/workflows/cdktf-deploy.yml`.

Reviewer: ```suggestion
Set the `TERRAFORM_CLOUD_TOKEN` environment variable to the [Terraform Cloud API token](https://www.terraform.io/cloud-docs/users-teams-organizations/api-tokens). This allows the CDKTF stacks `RemoteBackend` to connect to Terraform Cloud.

You must create a job like this for every stack so that you can pass the stack name into `cdktf diff <stack name>`. None of the stacks can depend on each other because Terraform cannot reason about which stacks must be launched first to respect these dependencies. 

The following example shows a GitHub Actions configuration where the CDKTF application is deployed from the `main` branch and uses Terraform Cloud. This file could be stored in `.github/workflows/cdktf-deploy.yml`.
```

      - name: Install Terraform
        uses: hashicorp/setup-terraform@v2
        with:
          terraform_version: 1.1.7
      - name: Install dependencies
        run: npm ci
      # If you use providers or modules that are not pre-built, you can install them here
      - name: Generate bindings for providers and modules
        run: cdktf get
      - name: Run Tests
        run: npm test
      - name: Deploy
        run: cdktf deploy --auto-approve '*' # Deploys all stacks
        env:
          TERRAFORM_CLOUD_TOKEN: ${{ secrets.TF_API_TOKEN }}
```

To get the plan output posted as a comment on your pull request, add the following to your `.github/workflows/cdktf-diff.yml` file:

```yaml
name: CDKTF Diff

on: [pull_request]

jobs:
  diff:
    name: "Terraform CDK Diff"
    runs-on: ubuntu-latest
    steps:
      - name: Checkout
        uses: actions/checkout@v2

      - name: Setup Terraform
        uses: hashicorp/setup-terraform@v1
        with:
          terraform_version: 1.1.7
          cli_config_credentials_token: ${{ secrets.TF_CLOUD_TOKEN }}

      - uses: actions/setup-node@v1
        with:
          node-version: "14"

      - run: yarn install

>       # We run the plan here, you need to fill in the name of your stack

Reviewer: ```suggestion
      # CDKTF runs the plan here, so you need to fill in the name of your stack.
```

      - name: plan
        run: cdktf plan <stack name>
        env:
          TERRAFORM_CLOUD_TOKEN: ${{ secrets.TF_API_TOKEN }}
        continue-on-error: true # We want to see the plan output even if the plan fails

>       # You can replace this step with any workflow you like, e.g. posting a message in a slack channel

Reviewer: ```suggestion
      # Replace this step with any workflow you like, e.g. posting a message in a slack channel.
```

      - uses: actions/github-script@0.9.0
        if: github.event_name == 'pull_request'
        env:

>           PLAN: "terraform\n${{ steps.plan.outputs.stdout }}" # We get the output from the last step

Reviewer: We generally want to avoid using "we" in documentation. Is this CDKTF? If it's actually Terraform, please correct me!

```suggestion
          PLAN: "terraform\n${{ steps.plan.outputs.stdout }}" # CDKTF gets the output from the last step
```

        with:

>           github-token: ${{ secrets.GH_COMMENT_TOKEN }} # You need to create a secret with a github token that can comment on PRs

Reviewer: ```suggestion
          github-token: ${{ secrets.GH_COMMENT_TOKEN }} # Create a secret with a GitHub token that can comment on PRs
```

          script: |
            const output = `#### Terraform Plan 📖\`${{ steps.plan.outcome }}\`
            <details><summary>Show Plan</summary>
            \`\`\`${process.env.PLAN}\`\`\`
            </details>
            *Pusher: @${{ github.actor }}, Action: \`${{ github.event_name }}\`, Working Directory: \`${{ env.tf_actions_working_dir }}\`, Workflow: \`${{ github.workflow }}\`*`;
            github.issues.createComment({
              issue_number: context.issue.number,
              owner: context.repo.owner,
              repo: context.repo.repo,
              body: output
            })
```

> Please be aware that you need a job like this for every stack you have as you have to pass the stack name into `cdktf diff <stack name>`. None of the stacks can be dependent on another one since you can only reliably plan a stack if the stack(s) it depends on are already deployed in their latest version. This is also the reason why `cdktf diff` only support a single stack and not multiple like `cdktf deploy` or `cdktf destroy` do.

Reviewer: I thin kthis should come before the examples so users can make an informed decision sooner about whether this workflow meets their use case.
```suggestion
```

### Running the CDKTF CLI on a general purpose CI

> If you have a general purpose CI like GitHub Actions or CircleCI setup already you can use the `cdktf deploy --auto-approve <your stacks>` command to trigger a deployment. This will run the deployment of your stacks in the right order and in parallel.
> 
> To set this up you need to have Node in version 14.0+ and the `cdktf-cli` CLI installed. If you are using a language that is not Node you need to install it as well. You also need to have Terraform installed, at least in version `1.0`.

Reviewer: ```suggestion
If you have a general purpose CI like GitHub Actions or CircleCI configured, use the `cdktf deploy --auto-approve <your stacks>` command to trigger a deployment. This deploys your stacks in the right order and in parallel. To do set up this workflow, you must install the following software:

 - Node in version 14.0+ 
 - The `cdktf-cli` CLI
 - Terraform v1.0+
 - Any other languages you use for your CDKTF application

~> **Warning:** Auto-approved deployments skip the option to review planned changes to your infrastructure before Terraform applies them. Minimize risk of unpredictable infrastructure changes and configuration drift by making sure that no one can change your infrastructure outside of your automated build pipeline.  Refer to [Non-Interactive Workflows](/cloud-docs/run/cli#non-interactive-workflows) for details.
```

> ### My CI solution / workflow is missing

Reviewer: ```suggestion
### Contribute Additional Examples
```

> We are continuously improving this page. If you have a solution that is missing, please [open an issue](https://github.com/hashicorp/terraform-cdk/issues/new?assignees=&labels=enhancement%2C+new&template=feature-request.md&title=). We are also happy to accept pull requests to extend this page.

Reviewer: ```suggestion
If you want to share another solution on this page, please [open a GitHub issue](https://github.com/hashicorp/terraform-cdk/issues/new?assignees=&labels=enhancement%2C+new&template=feature-request.md&title=) or submit a pull request for this page.
```
